### PR TITLE
General improvements to "Damaging Hits" section and armour breakdown.

### LINKED ISSUE #4240: PoB should account for spell suppression in "Maximum Hit Taken" if suppression chance is 100%
### Check for duplicates

- [X] I've checked for duplicate issues by using the search function of the [issue tracker](https://github.com/PathOfBuildingCommunity/PathOfBuilding/issues)

### Is your feature request related to a problem?

Currently, it seems that Spell Suppression is treated like dodge, block, or evasion in that it only increases your "effective hp pool" but not your maximum hit taken. But unlike those other defensive layers, suppression *can* reach 100% chance

### Describe the solution you'd like

Add a special case that checks if suppression chance is >=100%, and applies the suppression multiplier in max-hit-taken if so.

### Describe alternatives you've considered

_No response_

### Additional context

Wasn't sure if this is a "bug" or a "feature request", so I erred on the side of caution.

### PR DESCRIPTION
**Pull request by QuickStick123**
Fixes #4240.

### Description of the problem being solved:
The calculations for armour mitigations are very vague/confusing and have minimal breakdowns.
Spell suppression isn't included if it is guaranteed against spells.
Certain breakdown sections don't hide when irrelevant.
Generalised and separate resistances and base damage reduction. This makes it much easier to handle large changes to this in the future and allowed explicit references to damage types to be removed.

### Before Transcendence Armour Stacker
image
image
image
image
image
### After 
image
image
image
image
image

### Before Regular Build
image
image
### After 
image
image

### Before Enduring Cry Transcendence Without
image
image
### After
image
image

### Before Enduring Cry Transcendence With
image
### After
image

With 100% chance spell suppresion will appear in breakdown and get removed from other eHP calculations.
### Spell Suppression "Spell"
image

### Spell Suppression "Average"
image

### Link to a build that showcases this PR:
Transcendence Armour Stacker + Phys to ele https://pobb.in/S8lGNPQF8Af8
Regular Build https://pobb.in/3XXONK_WLL2U
Enduring cry transcendence https://pobb.in/h5VqofF3ymJH

### CODE REVIEW

File: src/Modules/CalcDefence.lua [hunk 1]
```diff
@@ -1073,84 +1073,108 @@ function calcs.defence(env, actor)
...
+					t_insert(breakdown[damageType.."DamageReduction"], s_format("Base Physical Damage Reduction: %d%%", resist))
+					if armourReduct ~= 0 then
+						t_insert(breakdown[damageType.."DamageReduction"], s_format("Total Physical Damage Reduction: %d%%", m_max(m_min(output.DamageReductionMax, resist - enemyPen + armourReduct), 0)))
+					end
 				end
-				t_insert(breakdown[damageType.."DamageReduction"], s_format("Reduction from Armour: %d%%", armourReduct))
 			end
 		end
 		local takenMult = output[damageType.."TakenHitMult"]
+		local spellSupressMult = 1
```
Comment (Nightblade) [inline src/Modules/CalcDefence.lua:1117]: `supress` should be `suppress` in all your changes to this file

File: src/Modules/CalcDefence.lua [hunk 2]
```diff
@@ -1071,86 +1073,117 @@ function calcs.defence(env, actor)
...
-					armourReduct = m_max(m_min(output.DamageReductionMax, resist - enemyPen + armourReduct), 0)
-					resist = armourReduct
+		if (modDB:Flag(nil, "ArmourAppliesTo"..damageType.."DamageTaken")) and not modDB:Flag(nil, "ArmourDoesNotApplyTo"..damageType.."DamageTaken") then
+			armourReduct = calcs.armourReduction(output.Armour * (1 + output.ArmourDefense), damage * resMult)
+			armourReduct = m_min(output.DamageReductionMax, armourReduct)
+		end
+		reductMult = (1 - m_max(m_min(output.DamageReductionMax, armourReduct + reduction - enemyOverwhelm), 0) / 100)
+		output[damageType.."DamageReduction"] = 100 - reductMult * 100
+		if reductMult ~= 1 then
+			if breakdown and reductMult ~= 1 then
```
Comment (Wires77) [inline src/Modules/CalcDefence.lua:1096]: This reductMult is redundant

File: src/Modules/CalcDefence.lua [hunk 3]
```diff
@@ -1673,7 +1708,7 @@ function calcs.defence(env, actor)
 			if output.ShowBlockEffect then
 				t_insert(breakdown["ConfiguredDamageChance"], s_format("x %.2f ^8(block effect)", output.BlockEffect / 100))
 			end
-			if suppressionEffect > 0 then
+			if suppressionEffect < 1 then
```
Comment (Wires77) [inline src/Modules/CalcDefence.lua:1711]: Could future-proof this by checking if it's ~= 1

**Review by Wires77 — CHANGES_REQUESTED** [review]
I found a couple of minor improvements, but this definitely looks a lot clearer.  One thing that is outside of scope for this PR, but we'll eventually want to fix is that armour reductions should only be applying to attacks, not spells.

**Review by Wires77 — APPROVED** [review]

**Pull request merged**